### DISCUSSION TITLE
Paginator Not Working With My Twitter Account

### DISCUSSION THREAD

**Discussion by SakiSakiSakiSakiSaki** [tweepy/tweepy#2056] [Questions] [answered]
Hi,

I am attempting to get my liked tweets from my profile, however my profile is locked/privated. I have some working code that only returns 20 liked tweets from my profile.

```
import tweepy
import configparser
import pandas as pd
import json, requests, re, os

#Credentials
config = configparser.ConfigParser(interpolation=None)
config.read('config.ini')
api_key = config['twitter']['api_key']
api_key_secret = config['twitter']['api_key_secret']
access_token = config['twitter']['access_token']
access_token_secret = config['twitter']['access_token_secret']
twitter_id = config['twitter']['twitter_id']

bearer_token = config['twitter']['bearer_token']

# Auth
auth = tweepy.OAuthHandler(api_key, api_key_secret)
auth.set_access_token(access_token, access_token_secret)


#Calling API
api = tweepy.API(auth, parser=tweepy.parsers.JSONParser())
client = tweepy.Client(bearer_token)
grabbed_tweets = api.get_favorites(user_id = twitter_id)

#-----------------------------
tweet_id_list = []

for tweet in grabbed_tweets:
    tweet_id_list.append(tweet['extended_entities']['media'][0]['id'])
print(len(tweet_id_list), "Tweets") #20 Tweets
```

I'd like all of my liked tweets, so I tried using paginator, however it returns nothing for me, while returning 250 tweets for other public Twitter profiles.

This is NASA's account, returning 250 tweets.
```
tweet_id_list = []
nasa_twitter_id = 11348282
for tweet in tweepy.Paginator(client.get_liked_tweets, id = nasa_twitter_id, max_results = 100).flatten(limit = 250):
    tweet_id_list.append(tweet.id)
print(len(tweet_id_list), "Tweets") #250 Tweets
```
This is my account, returning 0 tweets.
```
tweet_id_list = []
for tweet in tweepy.Paginator(client.get_liked_tweets, id = twitter_id, max_results = 100).flatten(limit = 250):
    tweet_id_list.append(tweet.id)
print(len(tweet_id_list), "Tweets") #0 Tweets
```

Am I doing something wrong? How can I fix this? Thanks.

**Comment by Harmon758** [comment]
Is `bearer_token` from OAuth 2.0 Authorization Code Flow with PKCE?

**Comment by SakiSakiSakiSakiSaki** [comment]
> Is `bearer_token` from OAuth 2.0 Authorization Code Flow with PKCE?

I'm not sure what "Code Flow with PKCE" means exactly, however `bearer_token` was generated from the Twitter Developers page for this project.

![image](https://user-images.githubusercontent.com/41088844/215916737-4f90a79f-584c-457c-9053-46002e7a2ecf.png)

I would like to mention that I do have 2FA on my account.

**Comment by Harmon758** [comment, marked as answer]
Since the account in question is private, you need to authenticate as that user with [OAuth 2.0 Authorization Code Flow with PKCE](https://tweepy.readthedocs.io/en/v4.12.1/authentication.html#oauth-2-0-authorization-code-flow-with-pkce-user-context) if you want to use an OAuth 2.0 access token. Otherwise, an app-only bearer token will not have access to the account.

Alternatively, you can just use your OAuth 1.0a User Context credentials or `Cursor` with `API` (and that endpoint/method's `count` parameter).

**Reply by SakiSakiSakiSakiSaki** [reply]
> Since the account in question is private, you need to authenticate as that user with [OAuth 2.0 Authorization Code Flow with PKCE](https://tweepy.readthedocs.io/en/v4.12.1/authentication.html#oauth-2-0-authorization-code-flow-with-pkce-user-context) if you want to use an OAuth 2.0 access token. Otherwise, an app-only bearer token will not have access to the account.
> 
> Alternatively, you can just use your OAuth 1.0a User Context credentials or `Cursor` with `API` (and that endpoint/method's `count` parameter).

Thanks for the answer. I've followed the documentation you linked, but I just reached a roadblock. I'm not sure what to put in `access_token = oauth2_user_handler.fetch_token()`

My full code is as such,
```
import tweepy
import configparser
import pandas as pd
import json, requests, re, os

#Credentials
config = configparser.ConfigParser(interpolation=None)
config.read('config.ini')
api_key = config['twitter']['api_key']
api_key_secret = config['twitter']['api_key_secret']
access_token = config['twitter']['access_token']
access_token_secret = config['twitter']['access_token_secret']
twitter_id = config['twitter']['twitter_id']
bearer_token = config['twitter']['bearer_token']
client_id = config['twitter']['client_id']
client_id_secret = config['twitter']['client_id_secret']

# Auth
auth = tweepy.OAuthHandler(api_key, api_key_secret)
auth.set_access_token(access_token, access_token_secret)

oauth2_user_handler = tweepy.OAuth2UserHandler(
    client_id = client_id,
    redirect_uri=  'https://example.org',
    scope = ["tweet.read", "users.read", "bookmark.read"],
    client_secret = client_id_secret
)

#print(oauth2_user_handler.get_authorization_url()) #Authorized the app with my personal account

access_token = oauth2_user_handler.fetch_token(
    oauth2_user_handler.get_authorization_url() #Not sure about this
)

client = tweepy.Client(access_token)
```

Which leads to the exception: 
```
Exception has occurred: MissingCodeError
(missing_code) Missing code parameter in response.
```

Would like to add that I don't have a domain or redirect uri that I'm hosting my code from.

**Reply by Harmon758** [reply]
As that documentation indicates, you need to pass the authorization response URL that you were redirected to.

**Reply by SakiSakiSakiSakiSaki** [reply]
I did read that part in the documentation, however I wasn't entirely sure if that was talking about the output of `print(oauth2_user_handler.get_authorization_url())`, or the page I was redirected to what I clicked on the "Authorize" button.

---
I tried both, the first led to the error I mentioned above, 
```
access_token = oauth2_user_handler.fetch_token(
    oauth2_user_handler.get_authorization_url() #Not sure about this
)
```
Which leads to the exception:

```
Exception has occurred: MissingCodeError
(missing_code) Missing code parameter in response.
```
---
The second gave me a different kind of error,
```
access_token = oauth2_user_handler.fetch_token("https://example.org/?state=BLAHBLAHBLAH&code=BLAHBLAHBLAH")
```
The error being,
```
Exception has occurred: MismatchingStateError
(mismatching_state) CSRF Warning! State not equal in request and response.
```
---
Sorry, am I missing something? A third method perhaps?

**Reply by Harmon758** [reply]
The latter should be correct.
What's the full traceback and all the relevant code?

**Reply by SakiSakiSakiSakiSaki** [reply]
Not sure if it's a security hazard to share the authorization response url, let me know if you need it to diagnose the issue:

My code:
```PY
import tweepy
import configparser
import pandas as pd
import json, requests, re, os

#Credentials
config = configparser.ConfigParser(interpolation=None)
config.read('config.ini')
api_key = config['twitter']['api_key']
api_key_secret = config['twitter']['api_key_secret']
access_token = config['twitter']['access_token']
access_token_secret = config['twitter']['access_token_secret']
twitter_id = config['twitter']['twitter_id']
bearer_token = config['twitter']['bearer_token']
client_id = config['twitter']['client_id']
client_id_secret = config['twitter']['client_id_secret']

# Auth
auth = tweepy.OAuthHandler(api_key, api_key_secret)
auth.set_access_token(access_token, access_token_secret)

oauth2_user_handler = tweepy.OAuth2UserHandler(
    client_id = client_id,
    redirect_uri=  'https://example.org',
    scope = ["tweet.read", "users.read", "bookmark.read"],
    client_secret = client_id_secret
)

print(oauth2_user_handler.get_authorization_url()) #Authorized the app with my personal account

access_token = oauth2_user_handler.fetch_token("https://example.org/?state=BLAHBLAHBLAH&code=BLAHBLAHBLAH")

client = tweepy.Client(access_token)

#Calling API
api = tweepy.API(auth, parser=tweepy.parsers.JSONParser())
grabbed_tweets = api.get_favorites(user_id = twitter_id)

#-----------------------------
tweet_id_list = []

for tweet in grabbed_tweets:
    tweet_id_list.append(tweet['extended_entities']['media'][0]['id'])
print(len(tweet_id_list), "Tweets") #20 Tweets

#-----------------------------
tweet_id_list = []
nasa_twitter_id = 11348282
for tweet in tweepy.Paginator(client.get_liked_tweets, id = nasa_twitter_id, max_results = 100).flatten(limit = 250):
    tweet_id_list.append(tweet.id)
print(len(tweet_id_list), "Tweets") #250 Tweets
```

The error that VSC had displayed to me:
```RED
Exception has occurred: MismatchingStateError
(mismatching_state) CSRF Warning! State not equal in request and response.
  File "C:\Users\Tweepy Project.py", line 31, in <module>
    access_token = oauth2_user_handler.fetch_token(
oauthlib.oauth2.rfc6749.errors.MismatchingStateError: (mismatching_state) CSRF Warning! State not equal in request and response.
```

**Reply by Harmon758** [reply]
Are you running this as a script or line by line?

**Reply by SakiSakiSakiSakiSaki** [reply]
I'm programming this in VSC, as well as running it through VSC, so I suppose line by line. I don't have a .py file I'm running through CMD or anything.

I guess the breakpoints I have are also considered line by line.

**Reply by Harmon758** [reply]
Is the URL you're passing the full URL you were redirected to by the preceding call to `get_authorization_url`?

**Reply by SakiSakiSakiSakiSaki** [reply]
In my actual code yes, I copied the full url from my browser and pasted it into my code.

**Reply by Harmon758** [reply]
As in you printed the authorization URL and then changed the next line with the call to `fetch_token` after authorizing with that specific URL, prior to executing that next line?

**Reply by SakiSakiSakiSakiSaki** [reply]
I think I understand what you're saying,

I paused my program after it printed the URL, change the `fetch_token` url to one that's just been authorized, and resumed the program.

However doing it that way led to an error, because I don't think it was properly updating.

I changed my code into include an `input` as a variable.

```
import tweepy
import configparser
import pandas as pd
import json, requests, re, os

#Credentials
config = configparser.ConfigParser(interpolation=None)
config.read('config.ini')
api_key = config['twitter']['api_key']
api_key_secret = config['twitter']['api_key_secret']
access_token = config['twitter']['access_token']
access_token_secret = config['twitter']['access_token_secret']
twitter_id = config['twitter']['twitter_id']
bearer_token = config['twitter']['bearer_token']
client_id = config['twitter']['client_id']
client_id_secret = config['twitter']['client_id_secret']

# Auth
auth = tweepy.OAuthHandler(api_key, api_key_secret)
auth.set_access_token(access_token, access_token_secret)

oauth2_user_handler = tweepy.OAuth2UserHandler(
    client_id = client_id,
    redirect_uri=  'https://example.org',
    scope = ["tweet.read", "users.read", "bookmark.read"],
    client_secret = client_id_secret
)

print(oauth2_user_handler.get_authorization_url()) #Authorized the app with my personal account

fetch_token_url = input('Please input the response url')

access_token = oauth2_user_handler.fetch_token(
    fetch_token_url
)

client = tweepy.Client(access_token)

#Calling API
api = tweepy.API(auth, parser=tweepy.parsers.JSONParser())
grabbed_tweets = api.get_favorites(user_id = twitter_id)

#-----------------------------
tweet_id_list = []

for tweet in grabbed_tweets:
    tweet_id_list.append(tweet['extended_entities']['media'][0]['id'])
print(len(tweet_id_list), "Tweets") #20 Tweets

#-----------------------------
tweet_id_list = []
nasa_twitter_id = 11348282
for tweet in tweepy.Paginator(client.get_liked_tweets, id = twitter_id, max_results = 100).flatten(limit = 250):
    tweet_id_list.append(tweet.id)
print(len(tweet_id_list), "Tweets") #250 Tweets
```

This managed to go through, however I was thrown another error when I tried to Paginator my profile.

```
Exception has occurred: Unauthorized
401 Unauthorized
Unauthorized
  File "C:\Users\Tweepy Project.py", line 53, in <module>
    for tweet in tweepy.Paginator(client.get_liked_tweets, id = twitter_id, max_results = 100).flatten(limit = 250):
tweepy.errors.Unauthorized: 401 Unauthorized
Unauthorized
```

**Reply by Harmon758** [reply]
You need to pass the specific authorization response URL that you're redirected to from the specific instance of the authorization URL that you're getting. Otherwise, there's a mismatch in the state, as the error indicates.

[GET /2/users/:id/liked_tweets](https://developer.twitter.com/en/docs/twitter-api/tweets/likes/api-reference/get-users-id-liked_tweets), the endpoint that [`Client.get_liked_tweets`](https://tweepy.readthedocs.io/en/v4.12.1/client.html#tweepy.Client.get_liked_tweets) uses, requires the `like.read` scope.

**Reply by SakiSakiSakiSakiSaki** [reply]
Do I have to get `auth_url`, authorize, get `auth_response_url`, every time I run this? Isn't that very cumbersome?

---
Would also like to mention that I still get the same `Unauthorized` exception after including the `like.read` scope,

```
import tweepy
import configparser
import pandas as pd
import json, requests, re, os

#Credentials
config = configparser.ConfigParser(interpolation=None)
config.read('config.ini')
api_key = config['twitter']['api_key']
api_key_secret = config['twitter']['api_key_secret']
access_token = config['twitter']['access_token']
access_token_secret = config['twitter']['access_token_secret']
twitter_id = config['twitter']['twitter_id']
bearer_token = config['twitter']['bearer_token']
client_id = config['twitter']['client_id']
client_id_secret = config['twitter']['client_id_secret']

# Auth
auth = tweepy.OAuthHandler(api_key, api_key_secret)
auth.set_access_token(access_token, access_token_secret)

oauth2_user_handler = tweepy.OAuth2UserHandler(
    client_id = client_id,
    redirect_uri=  'https://example.org',
    scope = ["tweet.read", "users.read", "bookmark.read", "like.read"],
    client_secret = client_id_secret
)

print(oauth2_user_handler.get_authorization_url()) #Authorized the app with my personal account

fetch_token_url = input('Please input the response url')

access_token = oauth2_user_handler.fetch_token(
    fetch_token_url
)

client = tweepy.Client(access_token)

#Calling API
api = tweepy.API(auth, parser=tweepy.parsers.JSONParser())
grabbed_tweets = api.get_favorites(user_id = twitter_id)

#-----------------------------
tweet_id_list = []

for tweet in grabbed_tweets:
    tweet_id_list.append(tweet['extended_entities']['media'][0]['id'])
print(len(tweet_id_list), "Tweets") #20 Tweets

#-----------------------------
tweet_id_list = []
nasa_twitter_id = 11348282
for tweet in tweepy.Paginator(client.get_liked_tweets, id = twitter_id, max_results = 100).flatten(limit = 250):
    tweet_id_list.append(tweet.id)
print(len(tweet_id_list), "Tweets") #250 Tweets
```

**Reply by SakiSakiSakiSakiSaki** [reply]
> You need to pass the specific authorization response URL that you're redirected to from the specific instance of the authorization URL that you're getting. Otherwise, there's a mismatch in the state, as the error indicates.
> 
> [GET /2/users/:id/liked_tweets](https://developer.twitter.com/en/docs/twitter-api/tweets/likes/api-reference/get-users-id-liked_tweets), the endpoint that [`Client.get_liked_tweets`](https://tweepy.readthedocs.io/en/v4.12.1/client.html#tweepy.Client.get_liked_tweets) uses, requires the `like.read` scope.

Just realized that not only am I getting that 401 error when I tried Paginator on my own Twitter ID, but I also get it when I use it on any Twitter ID. In the below code I tried NASA's twitter ID.
```PY
import tweepy
import configparser
import pandas as pd
import json, requests, re, os

#Credentials
config = configparser.ConfigParser(interpolation=None)
config.read('config.ini')
api_key = config['twitter']['api_key']
api_key_secret = config['twitter']['api_key_secret']
access_token = config['twitter']['access_token']
access_token_secret = config['twitter']['access_token_secret']
twitter_id = config['twitter']['twitter_id']
bearer_token = config['twitter']['bearer_token']
client_id = config['twitter']['client_id']
client_id_secret = config['twitter']['client_id_secret']

# Auth
auth = tweepy.OAuthHandler(api_key, api_key_secret)
auth.set_access_token(access_token, access_token_secret)

oauth2_user_handler = tweepy.OAuth2UserHandler(
    client_id = client_id,
    redirect_uri = 'https://example.org',
    scope = ["tweet.read", "users.read", "bookmark.read", "like.read"],
    client_secret = client_id_secret
)

print(oauth2_user_handler.get_authorization_url()) #Authorized the app with my personal account

fetch_token_url = input('Please input the response url: ')

access_token = oauth2_user_handler.fetch_token(
    fetch_token_url
)

client = tweepy.Client(access_token)

#Calling API
api = tweepy.API(auth, parser=tweepy.parsers.JSONParser())
grabbed_tweets = api.get_favorites(user_id = twitter_id)

#-----------------------------
tweet_id_list = []

for tweet in grabbed_tweets:
    tweet_id_list.append(tweet['extended_entities']['media'][0]['id'])
print(len(tweet_id_list), "Tweets") #20 Tweets

#-----------------------------
tweet_id_list = []
nasa_twitter_id = 11348282
for tweet in tweepy.Paginator(client.get_liked_tweets, id = nasa_twitter_id, max_results = 100).flatten(limit = 250):
    tweet_id_list.append(tweet.id)
print(len(tweet_id_list), "Tweets") #250 Tweets
```
Leading to the following error,
```RED
Exception has occurred: Unauthorized
401 Unauthorized
Unauthorized
  File "C:\Users\Tweepy Project.py", line 53, in <module>
    for tweet in tweepy.Paginator(client.get_liked_tweets, id = nasa_twitter_id, max_results = 100).flatten(limit = 250):
tweepy.errors.Unauthorized: 401 Unauthorized
Unauthorized
```
---

To illustrate what it is I am doing, I posted the below picture,

![Untitled-1](https://user-images.githubusercontent.com/41088844/216191169-7cb7019f-3572-4a39-9597-ce83d12843b0.png)
---
Not sure what's going on, would love some advice.

**Reply by Harmon758** [reply]
I'm not sure what you're asking. This is how you've chosen to write your code.

Are you using an access token that you've fetched with your updated scopes?

**Reply by SakiSakiSakiSakiSaki** [reply]
Yes, I've regenerated all of my tokens and keys and included them in my config.

What I'm asking is if there's an obvious reason why I'm getting a 401 error.

**Reply by Harmon758** [reply]
Oh, you need to pass the actual access token when initializing `Client`, not the dictionary that `fetch_token` returns.

**Reply by SakiSakiSakiSakiSaki** [reply]
I'm so dumb. Thanks. Works now.

---

> I'm not sure what you're asking. This is how you've chosen to write your code.

I think this question was in response to my question earlier,
> "Do I have to get auth_url, authorize, get auth_response_url, every time I run this? Isn't that very cumbersome?"

I think a better way of rephrasing would be, **is there way I can just authorize the app once, and never have to authorize it again?** 

_Another_ way of rephrasing, **does the `access_token` I saved, remain active? I noticed `'expires_in': 7200` and `'expires_at': 1675303070.1298165` elements in the dict we just spoke about.**

---

Apologies for the bad code, I'm not programmer lol

**Reply by Harmon758** [reply]
Not with OAuth 2.0 Authorization Code Flow with PKCE, no.

By default, OAuth 2.0 Authorization Code Flow with PKCE access tokens expire after two hours.
https://developer.twitter.com/en/docs/authentication/oauth-2-0/authorization-code

I believe refresh tokens expire after six months right now.

**Reply by SakiSakiSakiSakiSaki** [reply]
Is it as simple as including the `offline.access` scope?

Will the `access_token` included in that dict be a `refresh_token` instead? Same key name? Different?

**Reply by Harmon758** [reply]
I'm not sure what you're asking.

**Reply by SakiSakiSakiSakiSaki** [reply]
The documentation you linked mentions the following:

> If the scope `offline.access` is applied an OAuth 2.0 refresh token will be issued. With this refresh token, you obtain an access token. If this scope is not passed, we will not generate a refresh token.

What I’m asking is, if it is as simple as applying that scope. Do I need to do anything else? 

And if it is that simple, (I made a typo earlier, on mobile), will the dictionary that `fetch_token` outputs, contain said `refresh_token`? Will it be called `refresh_token`? Or will it still be called `access_token`?

**Reply by Harmon758** [reply]
To get a refresh token? Yes.
No.

Yes, you can simply test it out for yourself and see.
Yes.
No, the refresh token is not an access token.

**Reply by SakiSakiSakiSakiSaki** [reply]
> Yes, you can simply test it out for yourself and see.

Sorry was away from my machine, I tried it now, and got the `refresh_token`. Where am I supposed to pass this?

I tried the following and got a `401 Unauthorized` warning:

```PY
oauth2_user_handler = tweepy.OAuth2UserHandler(
    client_id = client_id,
    redirect_uri = 'https://example.org',
    scope = ["tweet.read", "users.read", "bookmark.read", "like.read", "offline.access"],
    client_secret = client_id_secret,
)

print(oauth2_user_handler.get_authorization_url()) #Authorized the app with my personal account
fetch_token_url = input('Please input the response url: ')
url_access_token_dict = oauth2_user_handler.fetch_token(
    fetch_token_url
)

client = tweepy.Client(url_access_token_dict['refresh_token'])
```
---
Reading the Tweepy [documentation](https://docs.tweepy.org/en/stable/authentication.html#:~:text=Auto_refresh_url,your%20access%20tokens.), I read:
> Auto_refresh_url
Refresh token endpoint URL, must be HTTPS. Supply this if you wish the client to automatically refresh your access tokens.

Is this what I'm missing in `oauth2_user_handler`? I tried various ways of implementing the `fetch_token_url` into `oauth2_user_handler`, but I couldn't get it right without an error.

I really appreciate you walking me through this.

**Reply by Harmon758** [reply]
As I said, the refresh token is not an access token.. This and how to use a refresh token are covered in that linked documentation.
What is your use case for this?

As for that parameter, it's a documentation mistake that I'm aware of and is due to `OAuth2UserHandler` subclassing `requests_oauthlib.OAuth2Session`.

**Reply by SakiSakiSakiSakiSaki** [reply]
> This and how to use a refresh token are covered in that linked documentation.

Ok so I see the following:
> classtweepy.OAuth2UserHandler(*, client_id, redirect_uri, scope, client_secret=None)
**token – Token dictionary, must include access_token and token_type.**

Let me know if I'm wrong, but is it something like,
```PY
oauth2_user_handler = tweepy.OAuth2UserHandler(
    client_id = client_id,
    redirect_uri = 'https://example.org',
    scope = ["tweet.read", "users.read", "bookmark.read", "like.read", "offline.access"],
    client_secret = client_id_secret,
    token = url_access_token_dict = oauth2_user_handler.fetch_token(fetch_token_url)
)
```

If not, could you tell me the proper syntax? 

---

>What is your use case for this?

My use case is downloading all of the images/videos I've liked on Twitter, and logging the metadata on a separate file. Before the API becomes Premium only.

If you mean specifically for `refresh_token`, I just find it cumbersome to have to do the app authorization and link copy/paste every two hours.

**Reply by Harmon758** [reply]
Again, that parameter is a mistake in the documentation due to `OAuth2UserHandler` subclassing `requests_oauthlib.OAuth2Session`. The signature for the class is correct. `Oauth2UserHandler` itself doesn't support refresh tokens on initialization yet.

Why do you need to use OAuth 2.0 Authorization Code Flow with PKCE for this?

**Reply by SakiSakiSakiSakiSaki** [reply]
> Again, that parameter is a mistake in the documentation due to `OAuth2UserHandler` subclassing `requests_oauthlib.OAuth2Session`. The signature for the class is correct. `Oauth2UserHandler` itself doesn't support refresh tokens on initialization yet.

So no `refresh_token` support yet for Tweepy, gotcha. Sorry! So is everyone (including you) just reauthorizing manually every 2 hours then?

---

> Why do you need to use OAuth 2.0 Authorization Code Flow with PKCE for this?

Earlier you mentioned the following:
>Since the account in question is private, **you need to authenticate as that user with [OAuth 2.0 Authorization Code Flow with PKCE]**(https://tweepy.readthedocs.io/en/v4.12.1/authentication.html#oauth-2-0-authorization-code-flow-with-pkce-user-context) if you want to use an OAuth 2.0 access token. Otherwise, an app-only bearer token will not have access to the account.

Again, my use case is:
> My use case is downloading all of the images/videos I've liked on Twitter, and logging the metadata on a separate file. Before the API becomes Premium only.

My account is private.

**Reply by Harmon758** [reply]
`requests_oauthlib.OAuth2Session` itself supports refresh tokens, and the Twitter API obviously does as well. There's countless other ways to make the necessary request to obtain an access token from a refresh token, and there's absolutely no issue with using those access tokens with Tweepy.

Yes, because you were attempting to use an OAuth 2.0 app-only bearer token.
If your code isn't user-facing, i.e. you're not going to be authenticating as any other user, and if you're not using any endpoints that require OAuth 2.0, why are you going through the entire OAuth flow instead of just using the OAuth 1.0a User Context credentials that you're already using? as I originally mentioned:
> Alternatively, you can just use your OAuth 1.0a User Context credentials

This entire discussion seems like a series of XY problems where I already provided you the simplest solution from the beginning.

**Reply by SakiSakiSakiSakiSaki** [reply]
> This entire discussion seems like a series of XY problems where I already provided you the simplest solution from the beginning.

I actually did try OAuth 1.0a when you suggested it, but I wasn't getting anything out of `client.get_liked_tweets()`.

```py
client = tweepy.Client(
    bearer_token=bearer_token,
    consumer_key=api_key,
    consumer_secret=api_key_secret,
    access_token=access_token,
    access_token_secret=access_token_secret,
    return_type = dict
)

tweet = client.get_liked_tweets(id = twitter_id)
print(tweet)
```

Returns
```
{'errors': [{'resource_id': 'MY TWITTER ID', 'parameter': 'id', 'resource_type': 'user', 'section': 'data', 'title': 'Authorization Error', 'value': 'MY TWITTER ID', 'detail': 'Sorry, you are not authorized to see the user with id: [MY TWITTER ID].', 'type': 'https://api.twitter.com/2/problems/not-authorized-for-resource'}]}
```

OAuth 2 did however, authorize me to view my account,
```
oauth2_user_handler = tweepy.OAuth2UserHandler(
    client_id = client_id,
    redirect_uri = 'https://example.org',
    scope = ["tweet.read", "users.read", "bookmark.read", "like.read", "offline.access"],
    client_secret = client_id_secret,

)

client = AuthorizeApp(oauth2_user_handler)
```
---
If anything, I'd rather use OAuth 1.0a as well. As you said, it's the simplest implementation in my code.
Am I missing something with OAuth 1.0a, to view my private profile and likes?

**Reply by Harmon758** [reply]
[`Client.get_liked_tweets`](https://tweepy.readthedocs.io/en/v4.12.1/client.html#tweepy.Client.get_liked_tweets) doesn't use OAuth 1.0a User Context by default, so it's using your bearer token.
You need to pass `True` for the `user_auth` parameter.

**Reply by SakiSakiSakiSakiSaki** [reply]
Alright that worked, thanks!
